src/lib/comms/sol-http-server-impl-microhttpd.c

```diff
@@ -0,0 +1,585 @@
+{
+    struct sol_http_request *request = data;
+    char *v, *k;
+
+    if (!size)
+        return MHD_NO;
+
+    v = strdup(value);
+    SOL_NULL_CHECK(v, MHD_NO);
+
+    k = strdup(key);
+    SOL_NULL_CHECK_GOTO(k, err);
+
+    if (!sol_http_param_add(&request->params,
+        SOL_HTTP_REQUEST_PARAM_POST_FIELD(k, v))) {
+        SOL_WRN("Could not add %s key", key);
+        goto param_err;
+    }
+
+    return MHD_YES;
+
+param_err:
+    free(k);
+err:
+    free(v);
+    return MHD_NO;
+}
+
+static time_t
+process_if_modified_since(const char *value)
+{
+    char *s;
+    struct tm t;
+
+    s = strptime(value, "%a, %d %b %Y %H:%M:%S GMT", &t);
+    if (!s)
+        return 0;
+
+    return timegm(&t);
+}
+
+static int
+headers_iterator(void *data, enum MHD_ValueKind kind, const char *key, const char *value)
+{
+    struct sol_http_request *request = data;
+    char *v, *k;
+
+    v = strdup(value);
+    SOL_NULL_CHECK(v, MHD_NO);
+
+    k = strdup(key);
+    SOL_NULL_CHECK_GOTO(k, err);
+
+    switch (kind) {
+    case MHD_HEADER_KIND:
+        if (!sol_http_param_add(&request->params,
+            SOL_HTTP_REQUEST_PARAM_HEADER(k, v)))
+            goto param_err;
+        if (streq(k, SOL_HTTP_PARAM_IF_SINCE_MODIFIED))
+            request->if_since_modified = process_if_modified_since(v);
+        break;
+    case MHD_COOKIE_KIND:
+        if (!sol_http_param_add(&request->params,
+            SOL_HTTP_REQUEST_PARAM_COOKIE(k, v)))
+            goto err;
+        break;
+    default:
+        goto param_err;
+    }
+
+    return MHD_YES;
+
+param_err:
+    free(k);
+err:
+    free(v);
+    return MHD_NO;
+}
+
+static int
+http_server_handler(void *data, struct MHD_Connection *connection, const char *url, const char *method,
+    const char *version, const char *upload_data, size_t *upload_data_size, void **ptr)
+{
+    int ret;
+    uint16_t i;
+    struct MHD_Response *mhd_response;
+    struct sol_http_server *server = data;
+    struct http_handler *handler;
+    struct sol_http_request *req = *ptr;
+    enum sol_http_status_code status = SOL_HTTP_STATUS_NOT_FOUND;
+
+    if (!req) {
+        req = sol_vector_append(&server->requests);
+        SOL_NULL_CHECK(req, MHD_NO);
+
+        sol_http_param_init(&req->params);
+        req->url = url;
+        req->connection = connection;
+        *ptr = req;
+        return MHD_YES;
+    }
+
+    if (method && streq(method, "POST")) {
+        req->method = SOL_HTTP_METHOD_POST;
+        if (!req->pp)
+            req->pp = MHD_create_post_processor(connection, 1024, post_iterator, req);
+        SOL_NULL_CHECK_GOTO(req->pp, end);
+
+        MHD_post_process(req->pp, upload_data, *upload_data_size);
+        if (*upload_data_size) {
+            *upload_data_size = 0;
+            return MHD_YES;
+        }
+    } else if (method && streq(method, "GET")) {
+        req->method = SOL_HTTP_METHOD_GET;
+    } else {
+        SOL_WRN("Method %s not implemented", method);
```

Comment:
If method is NULL an response with the status code 501 (not implemented) is sent.
I'm can change the warning to avoid %s with NULL.